rbd/mirror_nautilus.go

```diff
@@ -0,0 +1,379 @@
+	MirrorImageStatusStateSyncing = MirrorImageStatusState(C.MIRROR_IMAGE_STATUS_STATE_SYNCING)
+	// MirrorImageStatusStateStartingReplay is equivalent to MIRROR_IMAGE_STATUS_STATE_STARTING_REPLAY.
+	MirrorImageStatusStateStartingReplay = MirrorImageStatusState(C.MIRROR_IMAGE_STATUS_STATE_STARTING_REPLAY)
+	// MirrorImageStatusStateReplaying is equivalent to MIRROR_IMAGE_STATUS_STATE_REPLAYING.
+	MirrorImageStatusStateReplaying = MirrorImageStatusState(C.MIRROR_IMAGE_STATUS_STATE_REPLAYING)
+	// MirrorImageStatusStateStoppingReplay is equivalent to MIRROR_IMAGE_STATUS_STATE_STOPPING_REPLAY.
+	MirrorImageStatusStateStoppingReplay = MirrorImageStatusState(C.MIRROR_IMAGE_STATUS_STATE_STOPPING_REPLAY)
+	// MirrorImageStatusStateStopped is equivalent to MIRROR_IMAGE_STATUS_STATE_STOPPED.
+	MirrorImageStatusStateStopped = MirrorImageStatusState(C.MIRROR_IMAGE_STATUS_STATE_STOPPED)
+)
+
+// MirrorImageInfo provides information about the mirroring progress of an image.
+type MirrorImageInfo struct {
+	Name        string
+	Description string
+	State       MirrorImageState
+	StatusState MirrorImageStatusState
+	GlobalID    string
+	IsPrimary   bool
+	IsUp        bool
+}
+
+// MirrorGetImage returns the MirrorImageInfo for an image.
+//
+// Note: this can only be used if go-ceph is compiled with the `nautilus` build
+// tag.
+//
+// Implements:
+//  rbd_mirror_image_get_info(rbd_image_t image,
+//                            rbd_mirror_image_info_t *mirror_image_info,
+//                            size_t info_size)
+func (image *Image) MirrorGetImage() (*MirrorImageInfo, error) {
+	var status C.rbd_mirror_image_status_t
+
+	ret := C.rbd_mirror_image_get_status(image.image, &status, C.sizeof_rbd_mirror_image_status_t)
+	if ret != 0 {
+		return nil, getError(ret)
+	}
+
+	return &MirrorImageInfo{
+		Name:        C.GoString(status.name),
+		Description: C.GoString(status.description),
+		State:       MirrorImageState(status.info.state),
+		StatusState: MirrorImageStatusState(status.state),
+		GlobalID:    C.GoString(status.info.global_id),
+		IsPrimary:   bool(status.info.primary),
+		IsUp:        bool(status.up),
+	}, nil
+}
+
+// MirrorImageList returns a MirrorImageInfo for each mirrored image.
+//
+// Note: this can only be used if go-ceph is compiled with the `nautilus` build
+// tag.
+//
+// Implements:
+//  int rbd_mirror_image_status_list(rados_ioctx_t io_ctx,
+//  				      									 const char *start_id, size_t max,
+//  				      									 char **image_ids,
+//  				      									 rbd_mirror_image_status_t *images,
+//  				      									 size_t *len)
+func MirrorImageList(ioctx *rados.IOContext) ([]*MirrorImageInfo, error) {
+	mis := make([]*MirrorImageInfo, 0)
+
+	var imageListLen int
+
+	const cMaxIter C.size_t = 100
+
+	var cLen C.size_t
+	var startID string
+	cImageIDs := make([]*C.char, cMaxIter)
+	cImageStatus := make([]C.rbd_mirror_image_status_t, cMaxIter)
+
+	var ret C.int
+
+	for {
+		func() {
```

Comment:
Why not make this a top level private function? It'll be more explicit and save some indent levels. You can also move `var ret ...` into the function because I don't see it ever used outside the inner function's scope.

Reply:
I've made this change. Not sure the new version is any less ugly, but let me know what you think.

Reply:
Personally, I think the new version is better encapsulated. If I was writing the function, I'd have returned the next startID rather than use a pointer argument, but that's not worth changing unless you like that.
Also, I don't think you need imageListLen as a separate variable from cLen since Go will automatically compare numeric literals to C.size_t (IIRC). But again, change that only if you think it's worth it.

Reply:
I made the `cLen` change. Thanks for pointing that out.